theano/tensor/signal/pool.py

```diff
@@ -146,98 +220,82 @@ class Pool(OpenMPOp):
             interpreted as the number of rows, and the number of cols.
-        ds : list or tuple of two ints
-            Downsample factor over rows and columns this parameter indicates
-            the size of the pooling region.
-        st : list or tuple of two ints
-            The stride size. This is the distance between the pooling regions.
-            If it's set to None, it equals ds.
+        ds : list or tuple of N ints
+            Downsample factor over rows and column.
+            ds indicates the pool region size.
         ignore_border : bool
-            If ds doesn't divide imgshape, do we include an extra row/col of
-            partial downsampling (False) or ignore it (True).
-        padding : tuple of two ints
-            (pad_h, pad_w), pad zeros to extend beyond four borders
-            of the images, pad_h is the size of the top and bottom margins,
-            and pad_w is the size of the left and right margins.
+            If ds doesn't divide imgshape, do we include an extra row/col
+            of partial downsampling (False) or ignore it (True).
+        st : list or tuple of N ints or None
+            Stride size, which is the number of shifts over rows/cols/slices to get the
+            next pool region. If st is None, it is considered equal to ds
+            (no overlap on pooling regions).
+        padding: tuple of N ints or None
+            For each downsampling dimension, this specifies the number of zeros to
+            add as padding on both sides. No padding is added if padding is None.
 
         Returns
         -------
         list
             The shape of the output from this op, for input of given shape.
-            This will have the same length as imgshape, but with last two
+            This will have the same length as imgshape, but with last N
             elements reduced as per the downsampling & ignore_border flags.
 
         """
-        if len(imgshape) < 2:
-            raise TypeError('imgshape must have at least two elements '
-                            '(rows, cols)')
+        nd = len(ds)
+
+        if len(imgshape) < nd + 1:
+            raise TypeError('imgshape must have at least {} dimensions'.format(nd + 1))
 
         if st is None:
             st = ds
-        r, c = imgshape[-2:]
-        r = tensor.extract_constant(r)
-        c = tensor.extract_constant(c)
-        if padding[0]:
-            r += padding[0] * 2
-        if padding[1]:
-            c += padding[1] * 2
-
-        if ignore_border:
-            if ds[0] == st[0]:
-                nr = r // st[0]
-            else:
-                out_r = (r - ds[0]) // st[0] + 1
-                if isinstance(r, theano.Variable):
-                    nr = tensor.maximum(out_r, 0)
+        if padding is None:
+            padding = (0,) * nd
+        patch_shape = tuple(tensor.extract_constant(imgshape[-nd + i]) +
+                            (padding[i] * 2 if padding[i] else 0)
```

Comment:
Mainly because it was done like that in the original 2D code as well.
Looking at it again, I think that in the original version it has the advantage of not adding anything if the padding is 0. If the imgshape is a Theano variable, I can imagine that this is slightly more efficient because it doesn't add a + 0 operation. But yes, in my revised version this no longer holds. I'll remove it.